### PR TITLE
runtests.pl: kill processes locking test log files

### PR DESCRIPTION
**Pull request by mback2k**
For now only required and implemented for Windows.

Ref: #6058

### CODE REVIEW

**Comment by mback2k** [comment]
I will continue working on this. I also plan to implement Linux/macOS support for this via `lsof`.

**Comment by bagder** [comment]
on Linux and macOS the files aren't "locked" when another process is reading/writing them so what would the reason and idea be for this on those platforms?

**Comment by mback2k** [comment]
> on Linux and macOS the files aren't "locked" when another process is reading/writing them so what would the reason and idea be for this on those platforms?

That is correct, but it would still be good to know if the files are kept open while they should already be closed. So the detection part is probably a good idea, but maybe not the process killing part.

**Comment by mback2k** [comment]
@bagder you were right, since the deletion of "locked" files won't fail, the detection part wouldn't even run like on Windows.

**Comment by mback2k** [comment]
Rebased and ready for review.

File: tests/pathhelp.pm [hunk 1]
```diff
@@ -372,7 +372,15 @@ sub sys_native_abs_path {
         # Path is in relative form. Resolve relative directories in Unix form
         # *BEFORE* converting to Win32 form otherwise paths like
         # '../../../cygdrive/c/windows' will not be resolved.
-        my $cur_dir = `pwd -L`;
+
+        my $cur_dir;
+        # MSys shell has built-in command.
+        if($^O eq 'msys') {
+            $cur_dir = `bash -c 'pwd -L'`;
```
Comment (jay) [inline tests/pathhelp.pm:379]: why this

File: tests/runtests.pl [hunk 2]
```diff
@@ -2751,17 +2781,22 @@ sub cleardir {
...
+                if(!cleardir("$dir/$file")) {
+                    $done = 0;
+                }
+                if(!rmdir("$dir/$file")) {
+                    $done = 0;
+                }
             }
             else {
-                unlink("$dir/$file");
+                if(!unlink("$dir/$file") && "$file" !~ /_stunnel\.log$/) {
```
Comment (jay) [inline tests/runtests.pl:2792]: If you don't want to remove $file stunnel that check should happen before calling unlink.

File: tests/runtests.pl [hunk 3]
```diff
@@ -2737,12 +2737,42 @@ sub responsive_httptls_server {
...
+    if(pathhelp::os_is_win()) {
+        $dir = pathhelp::sys_native_abs_path($dir);
+        $dir =~ s/\//\\\\/g;
+        my $handle = "handle.exe";
+        if($ENV{"PROCESSOR_ARCHITECTURE"} =~ /64$/) {
+            $handle = "handle64.exe";
+        }
+        my @handles = `$handle $dir -accepteula -nobanner`;
+        for $handle (@handles) {
+            if($handle =~ /^(\S+)\s+pid:\s+(\d+)\s+type:\s+(\w+)\s+([0-9A-F]+):\s+(.+)\r\r/) {
```
Comment (jay) [inline tests/runtests.pl:2756]: why check for two carriage returns at the end, is this some weird quirk of msys i forgot about

**Review by jay — COMMENTED** [review]
I think you should make this behavior an option, like runtests.pl -f  forcibly kill lingering processes or something. This seems like a diagnostic for us to run in the CI and not for default use.

File: tests/pathhelp.pm (hunk 1 above)
Comment (mback2k) [reply tests/pathhelp.pm:379]: See commit description.

File: tests/runtests.pl (hunk 2 above)
Comment (mback2k) [reply tests/runtests.pl:2792]: I want to remove it, but ignore failure of that.

File: tests/runtests.pl (hunk 3 above)
Comment (mback2k) [reply tests/runtests.pl:2756]: That is specific to the output format of the handle tool.

**Comment by mback2k** [comment]
I consider it a general part for CI runs that should always be enabled, but yes: a flag for outside of CI makes sense.

File: tests/pathhelp.pm (hunk 1 above)
Comment (jay) [reply tests/pathhelp.pm:379]: ah I see that's perl specific. I can do `pwd -L` by itself from the shell but it doesn't work from perl. what about $ENV{PWD}

File: tests/runtests.pl (hunk 3 above)
Comment (jay) [reply tests/runtests.pl:2756]: I found something possibly related in some of my old code, if it's any help. It runs from msys perl to get service paths and grep to remove the \r\r and trailing whitespace and blank lines and header in wmic output
~~~perl
my @commands = grep { s/[ \t\r\n]*$//mg; !/^(PathName|)$/ }
               `wmic service GET PathName`;
~~~

File: tests/pathhelp.pm (hunk 1 above)
Comment (mback2k) [reply tests/pathhelp.pm:379]: That is not Perl specific, msys1 pwd binary just does not support -L so I use the one built into the shell. The environment variable does not return the path in the correct msys specific format/style and is dependent on the Perl interpreter.

File: tests/runtests.pl (hunk 3 above)
Comment (mback2k) [reply tests/runtests.pl:2756]: Thanks, I will remember this for future issues. I guess in this case the existing regex is just fine for the specific line parsed.

File: tests/runtests.pl (hunk 2 above)
Comment (mback2k) [reply tests/runtests.pl:2794]: Addressed with comment in newest commit.

**Review by jay — APPROVED** [review]
The commit message should say that this is a new option

**Pull request closed without merging**